### PR TITLE
Get dim from HNSW index

### PR DESCRIPTION
**Pull request by cydrain**
Signed-off-by: yudong.cai <yudong.cai@zilliz.com>

### CODE REVIEW

File: knowhere/index/vector_index/IndexHNSW.cpp [hunk 1]
```diff
@@ -252,7 +252,7 @@ IndexHNSW::QueryImpl(int64_t n,
         feder = std::make_unique<feder::hnsw::FederResult>();
     }
 
-    auto dim = GetMetaDim(config);
+    auto dim = index_->space_->get_data_size() / sizeof(float);
```
Comment (liliu-z) [inline knowhere/index/vector_index/IndexHNSW.cpp:255]: Why we don't pass dim in from `Query()` ?

**Review by liliu-z — COMMENTED** [review]
1. Can I ask why we make `QueryImp` and `RangeQueryImp` protected instead of private?
2. I noticed that we remove statistics related code. Can I ask what is the status of this features? If we need to remove them, suggest to remove them from all places(Knowhere & Milvus). If we only do this in HNSW, it brings consistency and confusion.

**Comment by cydrain** [comment]
> 1. Can I ask why we make `QueryImp` and `RangeQueryImp` protected instead of private?
QueryImpl() exists for many index types, such as IVF serial. For them, QueryImpl() is protected, so HNSW also mark it as protected.

> 2. I noticed that we remove statistics related code. Can I ask what is the status of this features? If we need to remove them, suggest to remove them from all places(Knowhere & Milvus). If we only do this in HNSW, it brings consistency and confusion.
Statistics related code is disabled since faiss upgraded from 1.6.3 to 1.7.2, I prefer to clean all of them from knowhere using another PR.

File: knowhere/index/vector_index/IndexHNSW.cpp (hunk 1 above)
Comment (cydrain) [reply knowhere/index/vector_index/IndexHNSW.cpp:255]: It's better to get dim from index

**Comment by liliu-z** [comment]
> > 1. Can I ask why we make `QueryImp` and `RangeQueryImp` protected instead of private?
> >    QueryImpl() exists for many index types, such as IVF serial. For them, QueryImpl() is protected, so HNSW also mark it as protected.
> 
> > 2. I noticed that we remove statistics related code. Can I ask what is the status of this features? If we need to remove them, suggest to remove them from all places(Knowhere & Milvus). If we only do this in HNSW, it brings consistency and confusion.
> >    Statistics related code is disabled since faiss upgraded from 1.6.3 to 1.7.2, I prefer to clean all of them from knowhere using another PR.

For the Statistics change, I am good to keep it what it is right now. We can set another PR to clean them all to make knowhere consistent on this.

For setting the QueryImp protected, this is not a good reason:
1. Protected in other indexes is due to inheritance purpose (Although it is not a good design, but implement is reasonable). And we don't have any inheritance here.
2. We need to mark all functions that will not get used outside this class as private.

File: knowhere/index/vector_index/IndexHNSW.cpp (hunk 1 above)
Comment (liliu-z) [reply knowhere/index/vector_index/IndexHNSW.cpp:255]: Then we have three place to get dim:
1. From `DataSet` passed in.
2. From `Dim()` function in `IndexHNSW`
3. From `index_->space_->get_data_size() / sizeof(float)`

this is too confusing, can we ensure they are consistent and which one should we pick?

**Pull request merged**